File: lenet_images/create_test_split.py
#!/usr/bin/env python
"""
Take a directory of ImageNet category directories and do a train / validation split
and write out train.txt and val.txt that create_database.sh needs
"""

import os
import sys
import yaml
import numpy as np
import pandas as pd
from sklearn.cross_validation import train_test_split

# --- CONFIGURE ---

# Read in system-specific configuration values. Edit this config.yaml file before running
config = yaml.load( open('config.yaml').read() )

# Make sure original image data directory is really a directory
if not os.path.isdir(config['test_image_data_root']):
    print
    print "Error: test_image_data_root is not a path to a directory:", config['test_image_data_root']
    print "Set the test_image_data_root variable in config.yaml to the path where the ImageNet training data is stored."
    sys.exit(1)


# --- RUN ---

print 'Reading image names and categories...'
# os.walk gives back a generator of tuples for each directory it walks
# ('current_dir', [directories], [files]), (...)
walk = os.walk(config['test_image_data_root'])
categories = walk.next()[1]

paths = []
cat_idxs = []

# Create list of image paths, with category directory as only prefix
for ii, category in enumerate(categories):
    file_list = walk.next()[2]
    n_files = len(file_list)
    paths.extend(['/'.join([category,filename]) for filename in file_list])
    cat_idxs.extend([ii]*n_files)


# Write out text files of file and label integer for both train and validation sets
print 'Writing\n  data/test.txt'
df = pd.DataFrame({'path':paths, 'label':[0]*len(cat_idxs)})
df[['path','label']].sort(['label','path']).to_csv('data/test.txt', sep=' ', header=None, index=None)

print '  data/test_answers.txt'
df = pd.DataFrame({'path':paths, 'label':cat_idxs})
df[['path','label']].sort(['label','path']).to_csv('data/test_answers.txt', sep=' ', header=None, index=None)

print '  data/test_categories.txt'
# Write out category names with their integer indices
df = pd.DataFrame(list(enumerate(categories)), columns=['idx','category'])
df.sort(['idx']).to_csv('data/test_categories.txt', sep='\t', header=None, index=None)

File: lenet_matlab/create_test_split.m
% Take a directory of ImageNet category directories and do a train / validation split
% and write out train.txt and val.txt that create_database.sh needs

%% --- CONFIGURE ---

% Read in system-specific configuration values. Edit this lenet_config.m file before running
config = lenet_config();

% Make sure original image data directory is really a directory
if ~isdir(config.train_image_data_root)
    fprintf(['Error: train_image_data_root is not a path to a directory: ', config.test_image_data_root', ' \n']);
    fprintf('Set the train_image_data_root variable in config.yaml to the path where the ImageNet training data is stored.\n');
    exit();
end

%% --- RUN ---

fprintf('Reading image names and categories...\n');

walk = dir(config.train_image_data_root);
is_dir = [walk(:).isdir];
categories = {walk(is_dir).name}';
categories(ismember(categories,{'.','..','.DS_Store'})) = [];

paths_test = [];
cat_dict = {};

% Create list of image paths, with category directory as only prefix
% concatenated with the category index
for ii = 1:length(categories),
    category = categories{ii};
    category_path = fullfile(config.train_image_data_root, category);
    walk = dir(category_path);
    is_dir = [walk(:).isdir];
    file_list = {walk(~is_dir).name}';
    n_files = length(file_list);
    for ff = 1:n_files,
        % category index 0-based
        file_list{ff} = [fullfile(category, file_list{ff}), ' ', int2str(ii-1)];
    end
    paths_test = [paths_test; sort(file_list)];
    cat_dict = [cat_dict; {[int2str(ii-1), ' ', category]}];
end

%% Write out text files of file and label integer for both train and validation sets
fprintf('Writing image names and categories...\n');

fid = fopen(fullfile('data','test.txt'), 'w');
for ii = 1:length(paths_test),
    fprintf(fid, '%s\n', paths_test{ii});
end
fclose(fid);

fid = fopen(fullfile('data','test_categories.txt'), 'w');
for ii = 1:length(cat_dict),
    fprintf(fid, '%s\n', cat_dict{ii});
end
fclose(fid);

